src/contracts/libraries/BN254.sol

```diff
@@ -131,6 +131,30 @@ library BN254 {
         require(success, "ec-add-failed");
     }
 
+    /**
+     * @notice an optimized ecMul implementation that takes O(log_2(s)) ecAdds
+     * @param p the point to multiply
+     * @param s the scalar to multiply by
+     * @dev this function is only safe to use if the scalar is 9 bits or less
```

Comment:
going to suggest here as well that we just do an explicit check for this and revert if the condition isn't met.
I'd rather the library be ~0.01% less gas-efficient than there be a single bug related to this